In [ ]:
NUM_RELU_R32 = 303104
NUM_RELU_R18 = 557056
NUM_RELU_MST = 10340

# Garbler Offline GC

I modified the [code](https://github.com/mc2-project/delphi/blob/master/rust/protocols/src/gc.rs#L169) to output the serialized size of the `ServerState` object.

Here is what I added:


```
....
println!("Len of encoders: {}", encoders.len());
println!("Len of output_randomizers: {}", output_randomizers.len());
let serialized_encoders: Vec<u8> = bincode::serialize(&encoders).unwrap();
let serialized_randomizers: Vec<u8> = bincode::serialize(&output_randomizers).unwrap();
let size_encoders = serialized_encoders.len()       //  
let size_randomizers = serialized_randomizers.len() // 
println!("Size of ServerState in bytes: {}, {}", size_encoders, size_randomizers);
Ok(ServerState{encoders,output_randomizers,})
....
```


Outputs:
- MNIST:
```
Len of encoders: 10340
Len of output_randomizers: 10340
Size of ServerState in Bytes: 37037888, 82728
```
- R32:
```
Len of encoders: 303104
Len of output_randomizers: 303104
Size of ServerState in Bytes: 1085718536, 2424840
```
- R18:
```
Len of encoders: 557056
Len of output_randomizers: 557056
Size of ServerState in Bytes: 1995374600, 4456456
```


In [ ]:
37037888 / NUM_RELU_MST , 82728 / NUM_RELU_MST

(3582.000773694391, 8.000773694390716)

In [ ]:
1085718536 / NUM_RELU_R32, 2424840 / NUM_RELU_R32

(3582.0000263935813, 8.00002639358108)

In [ ]:
1995374600 / NUM_RELU_R18, 4456456 / NUM_RELU_R18

(3582.0000143612133, 8.000014361213236)

In [ ]:
3582 + 8

3590


# Evaluator Offline GC

I modified the [code](https://github.com/mc2-project/delphi/blob/master/rust/protocols/src/gc.rs#L232) to output the serialized size of the `ClientState` object.

Here is what I added:

```
println!("Len of gc_s: {}", gc_s.len());
println!("Len of r_wires: {}", r_wires.len());
println!("Len of labels: {}", labels.len());
        
let encoded_gc_s: Vec<u8> = bincode::serialize(&gc_s).unwrap();
let encoded_r_wires: Vec<u8> = bincode::serialize(&r_wires).unwrap();
let encoded_labels: Vec<u8> = bincode::serialize(&labels).unwrap();

let size_gc = encoded_gc_s.len();
let size_rwires = encoded_r_wires.len();
let size_labels = encoded_labels.len();

println!("Size of ClientState in Bytes: {}, {}, {}", size_gc, size_rwires, size_labels);

Ok(ClientState {gc_s,server_randomizer_labels: r_wires, client_input_labels: labels,})
```

Outputs:
- MNIST:
```
Len of gc_s: 10340
Len of r_wires: 434280
Len of labels: 434280
Size of ClientState in Bytes: 164364648, 12159848, 12159848
```
- R32:
```
Len of gc_s: 303104
Len of r_wires: 12730368
Len of labels: 12730368
Size of ClientState in Bytes: 4818141192, 356450312, 356450312
```
- R18:
```
Len of gc_s: 557056
Len of r_wires: 23396352
Len of labels: 23396352
Size of ClientState in Bytes: 8854962184, 655097864, 655097864
```




In [ ]:
164364648/NUM_RELU_MST, 12159848/NUM_RELU_MST, 12159848/NUM_RELU_MST

(15896.00077369439, 1176.0007736943908, 1176.0007736943908)

In [ ]:
4818141192/NUM_RELU_R32, 356450312/NUM_RELU_R32, 356450312/NUM_RELU_R32

(15896.000026393582, 1176.000026393581, 1176.000026393581)

In [ ]:
8854962184/NUM_RELU_R18, 655097864/NUM_RELU_R18, 655097864/NUM_RELU_R18

(15896.000014361212, 1176.0000143612133, 1176.0000143612133)

In [ ]:
15896 + 2*1176

18248

# Overall Storage

For some network with some number of ReLUs..

The Garbler must store:

- 3590 Bytes / ReLU 

The Evaluator must store:

- 18248 Bytes / ReLU
